driver/src/price_estimation/dexag/api.rs

```diff
@@ -0,0 +1,167 @@
+            .timeout(timeout)
+            .build()
+            .context("failed to initialize HTTP client")?;
+        let base_url = base_url
+            .parse()
+            .with_context(|| format!("failed to parse url {}", base_url))?;
+        Ok(Self { base_url, client })
+    }
+}
+
+impl DexagApi for DexagApiImpl {
+    fn get_token_list(&self) -> Result<Vec<Token>> {
+        let mut url = self.base_url.clone();
+        url.set_path("token-list-full");
+        self.client
+            .get(url.to_string())
+            .context("failed to retrieve token list from dexag")?
+            .json::<Vec<Token>>()
+            .context("failed to parse token list json from dexag response")
+    }
+
+    fn get_price(&self, from: &Token, to: &Token, amount: EitherAmount) -> Result<Price> {
+        let mut url = self.base_url.clone();
+        url.set_path("price");
+        {
+            // This is in its own block because we are supposed to drop `q`.
+            let mut q = url.query_pairs_mut();
+            q.append_pair("from", &from.symbol);
+            q.append_pair("to", &to.symbol);
+            match amount {
+                EitherAmount::From(amount) => q.append_pair("fromAmount", &amount.to_string()),
+                EitherAmount::To(amount) => q.append_pair("toAmount", &amount.to_string()),
+            };
+            q.append_pair("dex", "ag");
+        }
+        self.client
+            .get(url.to_string())
+            .context("failed to retrieve price from dexag")?
+            .json::<Price>()
+            .context("failed to parse price json from dexag")
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    #[allow(clippy::float_cmp)]
+    fn deserialize_price() {
+        let json = r#"{"dex":"ag","price":"220.4","pair":{"base":"ETH","quote":"DAI"},"liquidity":{"uniswap":100}}"#;
+        let price: Price = serde_json::from_str(json).unwrap();
+        assert_eq!(price.price, 220.4);
+    }
+
+    #[test]
+    fn deserialize_token() {
+        use std::str::FromStr as _;
+        let json = r#"{"name":"SAI old DAI (SAI)","symbol":"SAI","address":"0x89d24a6b4ccb1b6faa2625fe562bdd9a23260359"}"#;
+        let token: Token = serde_json::from_str(json).unwrap();
+        assert_eq!(token.name, "SAI old DAI (SAI)");
+        assert_eq!(token.symbol, "SAI");
+        assert_eq!(
+            token.address.unwrap(),
+            Address::from_str("89d24a6b4ccb1b6faa2625fe562bdd9a23260359").unwrap()
+        );
+    }
+
+    #[test]
+    fn deserialize_token_no_address() {
+        let json = r#"{"name":"SAI old DAI (SAI)","symbol":"SAI"}"#;
+        let token: Token = serde_json::from_str(json).unwrap();
+        assert_eq!(token.name, "SAI old DAI (SAI)");
+        assert_eq!(token.symbol, "SAI");
+        assert!(token.address.is_none());
+    }
+
+    #[test]
+    #[ignore]
+    fn online_dexag_api() {
+        let api = DexagApiImpl::new().unwrap();
```

Comment:
Can you add a comment here with the command line needed to run this example? Just makes it easier for readers to run it and see the output.

Reply:
I don't like that we have this same comment for every one of these tests then but I'll add it for now.

Reply:
Any particular reason you’re not a fan of these comments?